lib/src/line_splitting/solve_state.dart

```diff
@@ -511,28 +508,29 @@ class SolveState {
         var value = _ruleValues.getValue(bound);
         var constraint = bound.constrain(value, unbound);
         if (constraint != null) {
-          _constraints[unbound] = constraint;
+          constraints[unbound] = constraint;
         }
       }
     }
+
+    _constraints = constraints;
+    return constraints;
   }
 
   /// Lazily initializes the [_unboundConstraints], which is needed to compare
   /// two states for overlap.
   ///
   /// We do this lazily because the calculation is a bit slow, and is only
   /// needed when we have two states with the same score.
-  void _ensureUnboundConstraints() {
-    if (_unboundConstraints != null) return;
-
-    // _ensureConstraints should be called first which initializes these.
-    assert(_boundRules != null);
-    assert(_unboundRules != null);
-
-    _unboundConstraints = {};
+  ///
+  /// [_ensureConstraints()] should be called first.
+  Map<Rule, Set<int>> _ensureUnboundConstraints() {
+    var unboundConstraints = _unboundConstraints;
+    if (unboundConstraints != null) return unboundConstraints;
 
+    unboundConstraints = <Rule, Set<int>>{};
     for (var unbound in _unboundRules) {
-      Set<int> disallowedValues;
+      Set<int>? disallowedValues;
```

Comment:
I'm tempted to suggest this:
```dart
late final Set<int> disallowedValues = unboundConstraints[unbound] = <int>{};
```

And then you can remove this code below:
```dart
if (disallowedValues == null) {
  disallowedValues = <int>{};
  unboundConstraints[unbound] = disallowedValues;
}
```
Basically you would be allowing lazy initialization to take care of storing `disallowedValues` in `unboundConstraints[unbound]`, only if there's actually at least one disallowed value.

But I dunno.  Maybe it's too clever to be comprehensible.  Your call.

Reply:
I dig it. Added a comment to make it a little clearer.